### PR TITLE
Converting set_merge to be set_intersect

### PR DESCRIPTION
**Pull request by johnlev**
Converted set_merge into set_intersect. This will be really helpful
 
 // Other minor tasks accomplished
 - Added intersect_helper_struct
- Modified set_merge and set_merge_helper
 
 ## Testing:
 Ran it

### CODE REVIEW

**Comment by Moyo97** [comment]
you don't know how to spell intersect :( #interEsect_helper smh

File: set_intersect.c [hunk 1]
```diff
@@ -67,33 +63,34 @@ int main()
 /* Merge the second set into the first set;
  * the second set is unchanged.
  */
-static void 
-set_merge(set_t *setA, set_t *setB)
+static set_t *set_intersect(set_t *setA, set_t *setB)
 {
-  set_iterate(setB, setA, set_merge_helper);
+  set_t *set = set_new();
```
Comment (mena-david) [inline set_intersect.c:71]: Should check inputs before using them

File: set_intersect.c [hunk 2]
```diff
@@ -67,33 +63,34 @@ int main()
...
+  set_intersect_data_t *intersect_data = assertp(malloc(sizeof(set_intersect_data_t)), "intersect_data");
+  
+  intersect_data->setA = setA;
+  intersect_data->setB = setB;
+  intersect_data->destination = set;
+
+  set_iterate(setA, intersect_data, set_intersect_helper);
+  set_iterate(setB, intersect_data, set_intersect_helper);
+  
+  return set;
```
Comment (mena-david) [inline set_intersect.c:82]: Memory Leaks! :( Please free intersect_data

File: set_intersect.c [hunk 3]
```diff
@@ -67,33 +63,34 @@ int main()
...
-    // not found: insert it
-    set_insert(setA, key, intsave(*itemB));
-    printf("\t%s added\n", key);
-  } else {
-    // add to the existing value
-    *itemA += *itemB;
-    printf("\t%s exists\n", key);
-  }
+  int *itemB = set_find(setB, key);
+  if (itemA != NULL && itemB != NULL)
```
Comment (uguryavuz) [inline set_intersect.c]: You may want to have this in brackets for better readability

Comment (mena-david) [inline set_intersect.c]: Need some good ol' Curly Braces Here. Please follow CS50 Guide Style.

File: set_intersect.c [hunk 4]
```diff
@@ -67,33 +63,34 @@ int main()
...
+    set_insert(destination, key, intsave(*itemB));
```
Comment (peter1357908) [inline set_intersect.c]: Why specifically choose itemB to save in the destination?

File: set_intersect.c [hunk 5]
```diff
@@ -67,33 +63,34 @@ int main()
...
+  set_iterate(setA, intersect_data, set_intersect_helper);
+  set_iterate(setB, intersect_data, set_intersect_helper);
+  
+  return set;
 }
 
-/* Consider one item for insertion into the other set.
- * If the other set does not contain the item, insert it;
- * otherwise, update the other set's item with sum of item values.
+/* Help the set_intersect function to complete it's opperation
```
Comment (peter1357908) [inline set_intersect.c]: Its*

File: set_intersect.c [hunk 6]
```diff
@@ -67,33 +63,34 @@ int main()
 /* Merge the second set into the first set;
```
Comment (uguryavuz) [inline set_intersect.c]: Update the documentation, this describes set_merge.

Comment (Moyo97) [inline set_intersect.c]: fix the comment here too. should say intersect :)

**Review by peter1357908 — CHANGES_REQUESTED** [review]
My comment on line 97 is ignored :(

File: set_intersect.c (hunk 4 above)
Comment (johnlev) [reply set_intersect.c]: Perhaps we take the min of the two?

Comment (peter1357908) [reply set_intersect.c]: K

**Review by mena-david — APPROVED** [review]
Good Job! I think we're ready to merge this!

**Review by peter1357908 — APPROVED** [review]

**Pull request merged**